+sw_tests/+unit_tests/unittest_spinw_optmagsteep.m

```diff
@@ -0,0 +1,223 @@
+            testCase.swobj.addmatrix('label', 'A', ...
+                                     'value', diag([0 0 -0.1])) % c easy
+            testCase.swobj.addmatrix('label', 'J1', 'value', 1)  % AFM
+            testCase.swobj.gencoupling();
+            testCase.swobj.addcoupling('mat', 'J1', 'bond', 1); % along a
+            testCase.swobj.addaniso('A');
+        end
+    end
+
+    methods (Test)
+        function test_no_nExt_throws_error(testCase)
+            testCase.swobj.genmagstr('mode', 'direct', 'S', [0; 0; 1], ...
+                                     'k',[0, 0, 0]);
+            func_call = @() testCase.swobj.optmagsteep('random', true);
+            testCase.verifyError(func_call, 'spinw:optmagsteep:NoField')
+        end
+        
+        function test_invalid_nExt_throws_error(testCase)
+            func_call = @() testCase.swobj.optmagsteep('nExt', [0, 0, 0]);
+            testCase.verifyError(func_call, 'spinw:optmagsteep:WrongInput')
+        end
+        
+        function test_warns_if_not_convereged(testCase)
+            % init moment along hard-axis (far from minima) and run 1 iter
+            testCase.swobj.genmagstr('mode', 'helical', 'S', [1; 0; 0], ...
+                                     'k',[0.5,0,0], 'n', [0,1,0], ...
+                                     'nExt', [2,1,1]);
+            func_call = @() testCase.swobj.optmagsteep('nRun', 1);
+            testCase.verifyWarning(func_call, ...
+                                   'spinw:optmagsteep:NotConverged')
+        end
+        
+        function test_warns_of_self_coupling_spins(testCase)
+            % add AFM exchange to bond along b (will double cell along b)
+            testCase.swobj.addmatrix('label', 'J2', 'value', 0.25)  % AFM
+            testCase.swobj.addcoupling('mat', 'J2', 'bond', 2); % along b
+            % run opt with k only along a
+            func_call = @() testCase.swobj.optmagsteep('NExt', [2, 1, 1], ...
+                                                       'nRun', 1);
+            testCase.verifyWarning(func_call, ...
+                                   'spinw:optmagsteep:SelfCoupling')
+        end
+        
+        function test_starting_from_groundstate_executes_one_iteration(testCase)
+            % generate ground state magnetic structure
+            testCase.swobj.genmagstr('mode', 'helical', 'S', [0; 0; -1], ...
+                                     'k',[0.5,0,0], 'n', [0,1,0], ...
+                                     'nExt', [2,1,1]);
+            opt = testCase.swobj.optmagsteep;
+            testCase.verify_val(testCase.swobj.magstr, ...
+                                testCase.default_magstr, 'abs_tol', 1e-6);
+            fields = {'datestart', 'dateend', 'obj', 'param'};
+            testCase.verify_val(rmfield(opt, fields), testCase.default_opt);
+            testCase.verify_obj(opt.obj, testCase.swobj)
+        end
+        
+        function test_converges_local_minimum_along_hard_axis(testCase)
```

Comment:
This is an interesting behaviour... it's expected from the algorithm but I'm not sure if it's something we want - e.g. even if the users give a hard axis starting value, when they call `optmagsteep` would they expect it to be refined towards the easy-axis? If so, should we add small random components to the initial spin directions before optimisation? (probably for another issue, not this PR... it might get complicated for more complex structures!)

Reply:
Yes I think the algorithm could try to guard against this - but at the end of the day you can always use random initial spin directions. It would be good if you could specify a random seed in the algorithm to make results reproducible (at least for testing - users should check the final state doesn't depend on the starting guess!)

Reply:
Yes, a good idea to set a random seed, but I think since it just uses `genmagstr` for the random spins directions, and [that in turn](https://github.com/SpinW/spinw/blob/080f3c3055538d4f878a706eb80a1aaaa05d69fe/swfiles/%40spinw/genmagstr.m#L394) uses the standard Matlab `randn` function, you can just use the `rng` function to set the seed - e.g. in the set up, add 

```
orig_rng_state = rng;
rng(12345); 
cleanup_rng = onCleanup(@()rng(orig_rng_state));
```

which should set the seed for the tests and then resets it after the test finishes and the `testCase` object is deleted.